activerecord/lib/active_record/reflection.rb

```diff
@@ -983,19 +983,18 @@ def add_as_through(seed)
         collect_join_reflections(seed + [self])
       end
 
-      def collect_join_reflections(seed)
-        a = source_reflection.add_as_source seed
-        if options[:source_type]
-          through_reflection.add_as_polymorphic_through self, a
-        else
-          through_reflection.add_as_through a
-        end
-      end
-
       private
+        def collect_join_reflections(seed)
+          a = source_reflection.add_as_source seed
+          if options[:source_type]
+            through_reflection.add_as_polymorphic_through self, a
+          else
+            through_reflection.add_as_through a
+          end
+        end
 
         def actual_source_reflection # FIXME: this is a horrible name
-          source_reflection.send(:actual_source_reflection)
+          source_reflection.actual_source_reflection
```

Comment:
> This isn't really private, so the change seems more misleading than explanatory.

Good point. Actually `actual_source_reflection` is called in `ThroughReflection`.
But `ThroughReflection` don't inherite `AssociationReflection`, so I moved `actual_source_reflection` from `AssociationReflection` to `AbstractReflection`.